Implement tunnels, elevated bridges, jumpjet and subterranean units.

pchote opened the pull request:
This PR introduces an `ICustomMovementLayer` interface and related plumbing that encode the idea of parallel movement/collision layers.  Mobile actors are able switch between these layers under implementation-defined conditions while moving.  This logic is then used for the core logic behind tunnels, jumpjets, and subterranean units.  An initial placeholder for elevated bridges is also included using a variation of tunnels combined with static prop actors.

The main priority here is the shared implementation and interaction with the pathfinder.  The TS features have their own sets of rough edges that I consider to be out of scope of this PR and can be improved on later.  I figure that this is a better approach than my usual long series of PRs to implement and polish everything in one pass.

pchote committed:
Add plumbing for custom movement layers.

pchote committed:
Implement tunnels.

pchote committed:
Fix CliffBackImpassibilityLayer trait name.

pchote committed:
Implement subterranean units.

pchote committed:
Hook up tunnels on Hot Springs.

pchote committed:
Fix units climbing tunnel faces.

pchote committed:
Disable actor sight while inside tunnels.

pchote committed:
Implement jumpjets.

pchote committed:
Add Condition support to WithDeathAnimation / DeathSounds.

pchote committed:
Implement elevated bridge placeholders.

pchote committed:
Add cell height display to map editor.

pchote committed:
Hook up bridges on Tread Lightly.

pchote committed:
Remove damaged bridge artwork from random tile selection.

pchote committed:
Hook up tunnels and elevated bridges in "A River Runs Near It".

pchote committed:
Hook up elevated bridges in Drawbridges.

pchote committed:
Move ActorMap to mod code.

pchote committed:
Remove CPos.Max/Min.

pchote committed:
Work around lint test limitations.

pchote commented:
Some of the issues that i've made note of:

Jumpjets:
* The jet animation stops when the actor is stopped mid-air.
* FallsToEarth can't trigger WithDeathAnimation, so corpses disappear once they hit the ground.
* Flying infantry don't have shadows because WithShadow doesn't work properly with upgrades.

Subterranean units:
* Cargo needs a condition to disable loading/unloading while underground.

Tunnels:
* The "enter" cursor on tunnel entrances to move to the other side isn't implemented.
* Units can hide inside tunnels where they become invulnerable.  Once the tunnel entrance code works idle units need to be able to automatically pick an entrance and leave.
* Cargo and other traits need condition support to disable them while inside tunnels.

Bridges:
* The whole implementation is a static placeholder similar to #11441, and needs to be rewritten after Mobile/DomainIndex have been refactored and we can support damage and repair.
* The input code needs to be able to distinguish between different movement layers to allow move orders onto elevated bridges.

I'll update this list with additional issues found during testing so that we can file tickets once this is merged.

GraionDilach commented:
I already bow down before you for the second commit it was on my next-month to-do list.

Gotta look into this when I can, notsure when that will happen though. >_>

Micr0Bit commented:
having the jumpjet standing on a closed plattform (with no entrance) and then doing a 2 shift-rightclicks to another plattform and back will make the unit teleport for a splitsecond out of the map ... 

i cant reproduce this always a 100%

here at 1:15 and 1:22

https://www.dropbox.com/s/o2ky5unj97dux3o/jetpack-spaz.orarep?dl=0

Phrohdoh commented:
At roughly 3:18 my `sapc`s are targeted and destroyed while being completely submerged:
https://drive.google.com/open?id=0B4Mag-oGBQNCTk1uaHRjU3ZuSHc

Replay was created on commit 6785b9ce8a811dec051802db6ccd30a911efbbd3

pchote committed:
Fix units warping to 0,0 when doing a 180 degree turn.

pchote commented:
Fixed both issues.  @Phrohdoh's was a silly oversight on my part (accidentally discarded important changes from a commit during a rebase), and @Micr0Bit's was a more general bug that i've fixed in a new commit.

RoosterDragon commented on OpenRA.Game/CPos.cs:
```diff
@@ -19,23 +19,22 @@ namespace OpenRA
-		public static CPos Max(CPos a, CPos b) { return new CPos(Math.Max(a.X, b.X), Math.Max(a.Y, b.Y)); }
-		public static CPos Min(CPos a, CPos b) { return new CPos(Math.Min(a.X, b.X), Math.Min(a.Y, b.Y)); }
-
 		public override int GetHashCode() { return X.GetHashCode() ^ Y.GetHashCode(); }
```
Consider hashing the layer too.

RoosterDragon commented on OpenRA.Game/CPos.cs:
```diff
@@ -19,23 +19,22 @@ namespace OpenRA
 	public struct CPos : IScriptBindable, ILuaAdditionBinding, ILuaSubtractionBinding, ILuaEqualityBinding, ILuaTableBinding, IEquatable<CPos>
 	{
 		public readonly int X, Y;
+		public readonly int Layer;
```
My gut reaction is worry that `CPos` is now another 4 bytes bigger and that's going to slow down other bits of code since we'll have to shuffle more bytes around for a fundamental type (particularly for maps that are now paying the cost for it despite not having extra layers).

Gut assumptions about performance are usually wrong, but until I can get some numbers one way or the other: How do you feel about `byte Layer` and possibly even `short X, Y` (hypothetically, for the moment)?

pchote replied:
I assume that the runtime will be aligning `CPos` instances on 32 or 64 bit boundaries, so making `Layer` a byte by itself shouldn't change anything... right?  We'd need to make X and Y shorts to win anything, but then it doesn't really matter what we make Layer (a byte would probably make sense though).

RoosterDragon replied:
Encoding `Layer` as a byte here means we can safely do the same elsewhere where it might actually save you something (if you had some collection of layer indexes, for.... something).

Besides, 256 layers ought to be enough for anybody.

RoosterDragon commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -440,10 +448,12 @@ protected override MovePart OnComplete(Actor self, Mobile mobile, Move parent)
 				}
 
+				var toPos = mobile.ToCell.Layer == 0 ? map.CenterOfCell(mobile.ToCell) :
+					((ActorMap)self.World.ActorMap).CustomMovementLayers[mobile.ToCell.Layer].CenterOfCell(mobile.ToCell);
```
You make this cast a lot to get at `CustomMovementLayers` - is there a reason to not just add that to the interface?

pchote replied:
`ActorMap` depends on `ICustomMovementLayer` which depends on `MobileInfo` (in Mods.Common), but `World.ActorMap` lives in Game.  This was the less bad alternative to querying MobileInfo from the actor for every pathfinder expand.

RoosterDragon replied:
That's fair.

RoosterDragon commented on OpenRA.Mods.Common/Traits/World/DomainIndex.cs:
```diff
@@ -37,8 +37,17 @@ public void WorldLoaded(World world, WorldRenderer wr)
+		public bool IsPassable(CPos p1, CPos p2, MobileInfo mi, uint movementClass)
 		{
+			// HACK: Work around units in other movement layers from being blocked
+			// when the poing in the main layer is not pathable
```
`poing` :P

RoosterDragon commented on OpenRA.Mods.Common/Activities/Move/MoveWithinRange.cs:
```diff
@@ -53,5 +54,18 @@ bool AtCorrectRange(WPos origin)
 			return Target.IsInRange(origin, maxRange) && !Target.IsInRange(origin, minRange);
 		}
+
+		bool InCorrectLayer(Actor self)
```
The duplication of this method would suggest `Target.IsInCorrectLayer` could be a good method to have.

pchote replied:
Damn... I was hoping nobody would notice this :P

pchote replied:
The Game/Mods split means we can't put this on Target itself, but it does make sense to deduplicate it somewhere.

RoosterDragon replied:
Somewhere, a blog post about how not to do separation of concerns is being written where we're every single reference. :)

pchote commented:
Thanks for the review! I'll get onto these on the weekend.

pchote commented:
Fixed @RoosterDragon's comments, rebased, and gave subterranean and jumpjet units speed boosts in their respective layers.

GraionDilach commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -73,6 +82,57 @@ public class MobileInfo : ConditionalTraitInfo, IMoveInfo, IPositionableInfo, IO
+		public readonly bool Subterranean = false;
+
+		[GrantedConditionReference]
+		[Desc("The condition to grant to self while inside a tunnel.")]
```
C'n'p much.

GraionDilach commented on OpenRA.Mods.Common/Traits/World/JumpjetActorLayer.cs:
```diff
@@ -0,0 +1,103 @@
+{
+	public class JumpjetActorLayerInfo : ITraitInfo
+	{
+		[Desc("Terrain type of the airbourne layer.")]
```
airbourne.

GraionDilach commented:
Gotta go over this tomorrow as well with testing and all that, but I found some typo nits.

pchote commented:
Fixed typos.

RoosterDragon commented:
I don't notice any obvious perf regressions in pathfinding, so no worries there.

pchote commented:
This is blocking several more PR's I'd like to open, and I know at least a few people have looked at this at least a bit, so it would be good if we can get some proper reviews in soon.

reaperrr commented on OpenRA.Mods.Common/Traits/World/TerrainTunnelLayer.cs:
```diff
@@ -0,0 +1,94 @@
+	public class TerrainTunnelLayerInfo : ITraitInfo, Requires<DomainIndexInfo>, ILobbyCustomRulesIgnore
+	{
+		[Desc("Terrain type used by cells outside any tunnel footprint.")]
+		public readonly string ImpassibleTerrainType = "Impassable";
```
I'm not a native english speaker, but using Impass**i**ble as field name but "Impassable" for the terrain type at least looks inconsistent.

reaperrr commented on mods/ts/sequences/misc.yaml:
```diff
@@ -609,7 +609,7 @@ tracks16:
 		ZRamp: 1
 		UseTilesetExtension: true
-		ShadowStart: 1
+		#ShadowStart: 1
```
I think this needs at least a short comment explaining that this should only be re-enabled when rendering has been refactored enough to fix shadow from hiding actors.

pchote replied:
This part of the tunnel shadow wasn't even used in the original game (at least on the version I tested, it might have been fixed or broken in a patch).

reaperrr replied:
You're right, nevermind then.

pchote replied:
Just wanted to point that out, i'm still going to comment it :smile:

reaperrr commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -82,6 +84,37 @@ public class MobileInfo : ConditionalTraitInfo, IMoveInfo, IPositionableInfo, IO
+		[Desc("The condition to grant to self while underground.")]
+		public readonly string SubterraneanCondition = null;
+
+		[Desc("Pathfinding cost for submerging or reemerging")]
```
Minor style nit: It would look better/be more consistent if you added a full stop at the end.

reaperrr commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -82,6 +84,37 @@ public class MobileInfo : ConditionalTraitInfo, IMoveInfo, IPositionableInfo, IO
+		public readonly string SubterraneanTransitionImage = null;
+
+		[SequenceReference("SubterraneanTransitionImage")]
+		[Desc("Dig animation image to play when transitioning")]
```
same here

reaperrr commented on mods/ts/rules/nod-vehicles.yaml:
```diff
@@ -293,12 +293,21 @@ SAPC:
 		RequiresCondition: !empdisable && !loading
+		Subterranean: true
+		SubterraneanCondition: submerged
+		SubterraneanTransitionTerrainTypes: Clear
```
You need to add `Rough` here, subterranean units were able to dig and emerge on the `Rough` terrain type in TS.

reaperrr commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -115,6 +116,22 @@ public class MobileInfo : ConditionalTraitInfo, IMoveInfo, IPositionableInfo, IO
+		[Desc("The condition to grant to self while flying.")]
+		public readonly string JumpjetCondition = null;
+
+		[Desc("Pathfinding cost for taking off or landing")]
```
Minor nit: Would look better with a full-stop.

reaperrr commented on OpenRA.Mods.Common/Traits/World/ElevatedBridgeLayer.cs:
```diff
@@ -0,0 +1,95 @@
+	public class ElevatedBridgeLayerInfo : ITraitInfo, Requires<DomainIndexInfo>, ILobbyCustomRulesIgnore
+	{
+		[Desc("Terrain type used by cells outside any elevated bridge footprint.")]
+		public readonly string ImpassibleTerrainType = "Impassable";
```
Minor nit: See my earlier comment about Impassible/Impassable.

reaperrr commented:
In addition to the minor things I pointed out, the jumpjet flight altitude seems to be a bit low (though hard to tell with no shadow).

Other than that, I didn't encounter any issues not listed in the OP, so this gets my 👍 once the comments have been adressed.

pchote commented:
Fixed nits and raised the jumpjet height from 2 to 2.25 cells to match the original game.

abcdefg30 commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -184,10 +184,15 @@ public override Activity Tick(Actor self)
 				var to = Util.BetweenCells(self.World, mobile.FromCell, mobile.ToCell) +
-					(self.World.Map.Grid.OffsetOfSubCell(mobile.FromSubCell) +
-					self.World.Map.Grid.OffsetOfSubCell(mobile.ToSubCell)) / 2;
+					(map.Grid.OffsetOfSubCell(mobile.FromSubCell) +	map.Grid.OffsetOfSubCell(mobile.ToSubCell)) / 2;
```
Tab after the `+`.

pchote replied:
Fixed.

reaperrr commented:
One question, is one of the follow-up by any chance about splitting locomotors from `Mobile`? Because in my opinion, `Mobile` was already too monolithic before, and with this it grows even more complex.

pchote commented:
Yes indeed, although I don't know when that will happen.

RoosterDragon reviewed (approved):
The code looks clean and pathfinding in the default mods doesn't seem adversely affected.

I won't pretend to be able to spot all the small issues in TS - but it makes to take this as-is and allow those to be worked out in time. I'm more the happy this passes the "feature worth having" and "doesn't break everything" bars - no need to wait on "works perfectly" for a non-shipping mod.

RoosterDragon commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/34e467a557169c655ba073062f69f6013bbd3d69)

GraionDilach commented:
I thought next release intends to ship TS and that's why the warning popup menu was added to the first place?

RoosterDragon commented:
Well in that case dock me some points for being out of the loop. If it really is on the horizon, we'll need this feature sooner rather than later though.

pchote commented:
I was originally planning to share a private/limited oramod release, but that is now off the table until the mod support stuff is fixed (which is blocked for the forseeable future on lack of reviews).  Including it as part of our main installer was never on the cards.

teees commented:
Great job, I just tested the tunnels and jumpjet troopers, works really well and way better than my initial approach 👍
